### PR TITLE
Add native video embedding support to Qwen3-VL

### PR DESCRIPTION
**Pull request by CharlesCNorton**
## Summary

- Add native video embedding support to Qwen3-VL via `model.mode` property
- Wire video path through standard `compute_embeddings` pipeline
- Auto-default mode from `dataset.media_type` when not explicitly set
- Add `video_fps` and `max_video_frames` config params

## Context

When `compute_embeddings` is called with an image model on a video dataset,
FiftyOne decomposes every video into individual frames and embeds each one
separately. For a dataset with dozens of videos and tens of thousands of
frames, this takes hours or days depending on hardware.

Qwen3-VL natively supports video input. It samples a handful of frames
internally and processes them as a single sequence. This PR adds a `mode`
property to the model so that `compute_embeddings` routes to the native
video path when `model.media_type == "video"`, producing one embedding
per video in a single forward pass.

Benchmarked on 3 videos (399 total frames) with RTX 6000 Ada:

| Path | Time | Rate |
|------|------|------|
| Per-frame (before) | 75.3s | 0.19s/frame, 399 forward passes |
| Native video (after) | 5.1s | 1.71s/video, 3 forward passes |

## Design

Behavior is uniquely defined by `model.media_type` and `dataset.media_type`:

| `model.media_type` | `dataset.media_type` | Behavior |
|---------------------|----------------------|----------|
| `"image"` | `"image"` | One embedding per image |
| `"image"` | `"video"` | Per-frame decomposition |
| `"video"` | `"video"` | One embedding per video (native) |

When `mode` is not explicitly set, it defaults to `dataset.media_type` at
inference time. A 3-line patch in `fiftyone/core/models.py:compute_embeddings`
handles this auto-detection.

## Usage

```python
import fiftyone as fo
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset("quickstart-video")

model = foz.load_zoo_model("qwen3-vl-embedding-2b-torch")

# mode auto-defaults to "video" for video datasets
dataset.compute_embeddings(model, embeddings_field="qwen_embs")

session = fo.launch_app(dataset)
```

Explicit mode control:

```python
# Native video embeddings (one per video)
model = foz.load_zoo_model("qwen3-vl-embedding-2b-torch", mode="video")
dataset.compute_embeddings(model, embeddings_field="video_embs")

# Per-frame decomposition (one per frame)
model.mode = "image"
dataset.compute_embeddings(model, embeddings_field="frame_embs")
```

Image datasets are unaffected. `mode` auto-defaults to `"image"`, and the
existing `compute_embeddings` path works as before.

## Config params

| Param | Default | Description |
|-------|---------|-------------|
| `mode` | `None` | `"image"` or `"video"`; auto-set from dataset if `None` |
| `video_fps` | `2.0` | Frame sampling rate for video inputs |
| `max_video_frames` | `128` | Cap on sampled frames; matches qwen-vl-utils `MAX_FRAMES` |

## Test plan

- [x] `compute_embeddings` mode=video on quickstart-video -- 3 videos, shape (3, 2048), 5.2s
- [x] `compute_embeddings` mode=image on quickstart -- 3 images, shape (3, 2048), 0.4s
- [x] Mode auto-default from dataset.media_type on both image and video datasets
- [x] Mode switching on same model instance
- [x] `compute_visualization` / `compute_similarity` / `sort_by_similarity` on video embeddings
- [x] Existing unit tests -- 36/36 pass

## Summary by CodeRabbit

* **New Features**
  * Video embedding support with configurable frame rate, max frames, and selectable mode (image/video/auto).
  * Automatic media-type detection so embedding adapts to image or video inputs.

* **Chores**
  * Unified embedding flow and standardized post-processing (consistent sizing and optional normalization).
  * Model mode temporarily aligned to input media type during embedding and reset afterward.
  * Dataset field handling adjusted to ensure correct sample- vs frame-level embedding storage.

* **Bug Fixes**
  * Ensure a defined embedding value on per-sample video embedding failures.

* **Tests**
  * Added unit tests for mode behavior, auto-mode detection, and video embedding.

### CODE REVIEW

File: fiftyone/utils/qwen3_vl.py [hunk 1]
```diff
@@ -217,8 +223,17 @@ class Qwen3VLModel(fout.TorchImageModel, fom.EmbeddingsMixin):
...
+        self._mode = config.mode
         super().__init__(config)
 
+    @property
+    def mode(self):
+        return self._mode
+
+    @mode.setter
+    def mode(self, value):
+        self._mode = value
```
[the remark this answers is not in the document]
Comment (lorenzoriano) [reply fiftyone/utils/qwen3_vl.py:261]: @CharlesCNorton agree it'll be better if a check was done here

File: fiftyone/utils/qwen3_vl.py [hunk 2]
```diff
@@ -383,14 +405,20 @@ def _prepare_image(self, img):
         return img
```
Comment (lorenzoriano) [inline fiftyone/utils/qwen3_vl.py:423]: <p class='graphite__hidden'><i>[Re: line +389]</i></p>
I don't think we benefit from an import inside a function
<p class='graphite__hidden'>See this comment inline on <a href="https://app.graphite.com/github/pr/voxel51/fiftyone/7049?utm_source=unchanged-line-comment">Graphite</a>.</p>

File: fiftyone/core/models.py [hunk 3]
```diff
@@ -1040,6 +1040,10 @@ def compute_embeddings(
```
Comment (lorenzoriano) [inline fiftyone/core/models.py:1045]: <p class='graphite__hidden'><i>[Re: line +1025]</i></p>
field registration happens before auto mode injection. When mode=None on a video dataset, it will first register a frame field, but then switch to video mode and write sample embeddings. 
<p class='graphite__hidden'>See this comment inline on <a href="https://app.graphite.com/github/pr/voxel51/fiftyone/7049?utm_source=unchanged-line-comment">Graphite</a>.</p>

File: fiftyone/core/models.py [hunk 4]
```diff
@@ -1040,6 +1040,10 @@ def compute_embeddings(
 
         context.enter_context(model)
 
+        if hasattr(model, "mode") and model.mode is None:
+            model.mode = samples.media_type
```
Comment (lorenzoriano) [inline fiftyone/core/models.py]: This should happen before before, around line 1012

File: fiftyone/utils/qwen3_vl.py [hunk 5]
```diff
@@ -403,3 +431,79 @@ def embed_all(self, args):
...
+            if i % step == 0:
+                frames.append(self._prepare_image(frame))
+                if len(frames) >= self.config.max_video_frames:
+                    break
+
+        if not frames:
+            return np.zeros(
+                self.config.embedding_dim or self._model.config.hidden_size,
+                dtype=np.float32,
+            )
```
Comment (lorenzoriano) [inline fiftyone/utils/qwen3_vl.py]: In this case the embeddings are not normalized. I don't know how to fix this, probably the best solution would be to raise an error, or skip the frames

File: fiftyone/utils/qwen3_vl.py [hunk 6]
```diff
@@ -403,3 +431,79 @@ def embed_all(self, args):
...
+
+        messages = [
+            {
+                "role": "user",
+                "content": [
+                    {
+                        "type": "video",
+                        "video": frames,
+                        "fps": sample_fps,
```
Comment (lorenzoriano) [inline fiftyone/utils/qwen3_vl.py]: when raw_fps is not a multiple of sample_fps (or if sample_fps > raw_fps) the FPS of sampled frames are wrong. Use  effective_fps = raw_fps / step

File: fiftyone/utils/qwen3_vl.py [hunk 7]
```diff
@@ -217,8 +230,17 @@ class Qwen3VLModel(fout.TorchImageModel, fom.EmbeddingsMixin):
 
     def __init__(self, config):
         self._processor = None
+        self._mode = config.mode
```
Comment (lorenzoriano) [inline fiftyone/utils/qwen3_vl.py:248]: It' d be better to check for valid modes, eg. `("image", "video", None)`

**Review by lorenzoriano — CHANGES_REQUESTED** [review]

File: fiftyone/utils/qwen3_vl.py (hunk 7 above)
Comment (CharlesCNorton) [reply fiftyone/utils/qwen3_vl.py:248]: Fixed in 2446b1b

File: fiftyone/utils/qwen3_vl.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (CharlesCNorton) [reply fiftyone/utils/qwen3_vl.py:261]: :+1:

File: fiftyone/utils/qwen3_vl.py (hunk 2 above)
Comment (CharlesCNorton) [reply fiftyone/utils/qwen3_vl.py:423]: Moved to module level in 70d4a6b.

File: fiftyone/core/models.py (hunk 3 above)
Comment (CharlesCNorton) [reply fiftyone/core/models.py:1045]: Fixed in 3214660 — moved auto-mode injection to top of the `with` block, before field registration.

File: fiftyone/core/models.py (hunk 4 above)
Comment (CharlesCNorton) [reply fiftyone/core/models.py]: :+1: Done in 3214660.

File: fiftyone/utils/qwen3_vl.py (hunk 5 above)
Comment (CharlesCNorton) [reply fiftyone/utils/qwen3_vl.py]: Now raises `ValueError` — fd7cde5.

File: fiftyone/utils/qwen3_vl.py (hunk 6 above)
Comment (CharlesCNorton) [reply fiftyone/utils/qwen3_vl.py]: Fixed in fd7cde5.

**Comment by CharlesCNorton** [comment]
0.0.14 is the only release with Qwen3-VL support and the fix ([PR #2045](https://github.com/QwenLM/Qwen3-VL/pull/2045)) hasn't shipped, so there's no version to pin around it. For embeddings the resolution difference is negligible. Will bump the pin when a fixed release lands.

File: fiftyone/utils/qwen3_vl.py [hunk 8]
```diff
@@ -349,18 +385,21 @@ def _embed_images(self, imgs):
...
 
-            embeddings.append(embedding.float().cpu().numpy())
+        if self.config.normalize_embeddings:
+            embedding = torch.nn.functional.normalize(embedding, p=2, dim=-1)
 
-        return np.vstack(embeddings)
+        return embedding.float().cpu().numpy()
 
     def _prepare_image(self, img):
         """Convert image to PIL format for processor."""
```
Comment (lorenzoriano) [inline fiftyone/utils/qwen3_vl.py:408]: <p class='graphite__hidden'><i>[Re: line +406]</i></p>
remove import
<p class='graphite__hidden'>See this comment inline on <a href="https://app.graphite.com/github/pr/voxel51/fiftyone/7049?utm_source=unchanged-line-comment">Graphite</a>.</p>

File: fiftyone/utils/qwen3_vl.py [hunk 9]
```diff
@@ -403,3 +447,83 @@ def embed_all(self, args):
...
+        message to Qwen3-VL, which processes the full temporal context
+        and returns one embedding vector.
+
+        Args:
+            video_reader: an ``eta.core.video.FFmpegVideoReader``
+
+        Returns:
+            a 1D numpy array embedding
+        """
+        from PIL import Image as PILImage
```
Comment (lorenzoriano) [inline fiftyone/utils/qwen3_vl.py]: remove import

File: fiftyone/utils/qwen3_vl.py [hunk 10]
```diff
@@ -245,7 +281,7 @@ def _load_model(self, config):
 
     @property
     def media_type(self):
-        return "image"
+        return self._mode or "image"
 
     def _get_prompt(self):
         if self.config.prompt is not None:
```
Comment (lorenzoriano) [inline fiftyone/utils/qwen3_vl.py:292]: <p class='graphite__hidden'><i>[Re: line +308]</i></p>
remove import
<p class='graphite__hidden'>See this comment inline on <a href="https://app.graphite.com/github/pr/voxel51/fiftyone/7049?utm_source=unchanged-line-comment">Graphite</a>.</p>

File: fiftyone/utils/qwen3_vl.py [hunk 11]
```diff
@@ -403,3 +447,83 @@ def embed_all(self, args):
...
+                    {
+                        "type": "video",
+                        "video": frames,
+                        "fps": effective_fps,
+                    },
+                ],
+            }
+        ]
+
+        from qwen_vl_utils import process_vision_info
```
Comment (lorenzoriano) [inline fiftyone/utils/qwen3_vl.py]: move import to module

File: tests/unittests/utils/test_qwen3_vl.py [hunk 12]
```diff
@@ -398,5 +400,167 @@ def test_prepare_image_hwc_small_height(self):
...
+        mock_reader = mock.MagicMock()
+        mock_reader.__iter__ = mock.Mock(return_value=iter([]))
+        with mock.patch(
+            "fiftyone.core.models.etav.FFmpegVideoReader",
+            return_value=mock_reader,
+        ):
+            mock_reader.__enter__ = mock.Mock(return_value=mock_reader)
+            mock_reader.__exit__ = mock.Mock(return_value=False)
+            ds.compute_embeddings(model, embeddings_field="emb")
+        assert ds.has_frame_field("emb")
```
Comment (lorenzoriano) [inline tests/unittests/utils/test_qwen3_vl.py:480]: I think this will always pass, since compute_embeddings will run add_frame_field. Any way this test can actually made more stressfull?

File: tests/unittests/utils/test_qwen3_vl.py [hunk 13]
```diff
@@ -398,5 +400,167 @@ def test_prepare_image_hwc_small_height(self):
...
+        assert not ds.has_sample_field("emb")
+        assert model.mode == "image"
+        ds.delete()
+
+    def test_mode_none_image_dataset(self):
+        """mode=None on image dataset -> sample-level embeddings"""
+        import PIL.Image
+        import tempfile
+        import os
```
Comment (lorenzoriano) [inline tests/unittests/utils/test_qwen3_vl.py]: I think we should instruct the AI bot always to have imports at the top of the file

File: fiftyone/utils/qwen3_vl.py [hunk 14]
```diff
@@ -25,9 +27,9 @@
 
 
 def _ensure_qwen3_vl():
-    fou.ensure_package("transformers>=4.51.0")
+    fou.ensure_package("transformers>=4.57.0")
     fou.ensure_package("accelerate")
-    fou.ensure_package("qwen-vl-utils")
+    fou.ensure_package("qwen-vl-utils>=0.0.1")
```
[the remark this answers is not in the document]
Comment (lorenzoriano) [reply fiftyone/utils/qwen3_vl.py:34]: If this is valid can you address this?

File: tests/unittests/utils/test_qwen3_vl.py [hunk 15]
```diff
@@ -398,5 +400,183 @@ def test_prepare_image_hwc_small_height(self):
...
+        model.mode = "video"
+        assert model.mode == "video"
+        assert model.media_type == "video"
+
+
+class TestQwen3VLAutoMode:
+    """Test that mode auto-defaults from dataset media type via compute_embeddings"""
+
+    def _make_mock_model(self):
+        import fiftyone.core.models as fom
```
Comment (lorenzoriano) [inline tests/unittests/utils/test_qwen3_vl.py]: import inside function (this is one example, there are several others in his file)

File: tests/unittests/utils/test_qwen3_vl.py [hunk 16]
```diff
@@ -398,5 +400,183 @@ def test_prepare_image_hwc_small_height(self):
...
+        model = self._make_mock_model()
+        tmp = os.path.join(tempfile.gettempdir(), "test_auto_mode.png")
+        PIL.Image.new("RGB", (10, 10)).save(tmp)
+        ds = fo.Dataset()
+        ds.add_sample(fo.Sample(filepath=tmp))
+        ds.compute_embeddings(model, embeddings_field="emb")
+        assert ds.has_sample_field("emb")
+        assert model.mode is None
+        ds.delete()
+        os.remove(tmp)
```
Comment (lorenzoriano) [inline tests/unittests/utils/test_qwen3_vl.py]: The other video tests correctly use the tmp_path fixture, it guarantees that it auto clean. Here it doesn't if the test fail. Can you change it here the same way as the others?

File: fiftyone/core/models.py [hunk 17]
```diff
@@ -1009,32 +1009,36 @@ def compute_embeddings(
...
-    )
-
-    use_data_loader = (
-        isinstance(model, (SupportsGetItem, TorchModelMixin))
-        and not process_video_frames
-    )
+    with contextlib.ExitStack() as context:
+        if hasattr(model, "mode") and model.mode is None:
+            model.mode = samples.media_type
+            context.callback(setattr, model, "mode", None)
```
Comment (lorenzoriano) [inline fiftyone/core/models.py]: nit: can you use `context.enter_context(fou.SetAttributes(model, preprocess=False))`like at line 1043?

File: tests/unittests/utils/test_qwen3_vl.py (hunk 12 above)
Comment (lorenzoriano) [reply tests/unittests/utils/test_qwen3_vl.py:480]: What I mean here, can you test instead:

- The correct number of embeddings were generated
- The embeddings have the expected shape
- The actual embedding values were computed (not just the field existence)

Comment (CharlesCNorton) [reply tests/unittests/utils/test_qwen3_vl.py:480]: Addressed in f64dcc4 👍
